change: ETCM-7811 native token inherent data provider

AmbientTea opened the pull request:
# Description

Adds the inherent data provider that uses the data source from the previous PR and produces inherent data containing the number of tokens moved into the illiquid supply. I wired it into the node already so the inherent data is already produced, but there is no pallet to use it yet. 

The next PR is [the pallet](https://github.com/input-output-hk/partner-chains/pull/36). You can see the changes from all the PR together in [this draft PR](https://github.com/input-output-hk/partner-chains/pull/22). 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages.
- [x] New tests are added if needed and existing tests are updated.
- [x] Relevant logging and metrics added
- [ ] CI passes. See note on CI.
- [ ] Any changes are noted in the `changelog.md` for affected crate
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

mpskowron commented on primitives/native-token-management/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+	#[derive(thiserror::Error, sp_runtime::RuntimeDebug)]
+	pub enum IDPCreationError {
+		#[error("Failed to read native token data from data source: {0:?}")]
+		DataSourceError(DataSourceError),
```
I think that you can use a @from annotation here instead of manually adding From implementation below.

mpskowron commented on primitives/native-token-management/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+				.map_err(IDPCreationError::GetMainChainScriptsError)?;
+			let parent_mc_hash: Option<McBlockHash> = client
+				.header(parent_hash)
+				.unwrap()
```
Why unwrap()?

mpskowron commented on primitives/native-token-management/src/lib.rs:
```diff
@@ -0,0 +1,127 @@
+				.unwrap()
+				.filter(|parent_header| !parent_header.number().is_zero())
+				.map(|parent_header| {
+					McHashInherentDigest::value_from_digest(&parent_header.digest().logs).unwrap()
```
Why unwrap()?

AmbientTea commented:
> Why unwrap()?

Right, I changed each error to be returned not panicked. I also moved the logic of extracting the mc hash from a header from the native token crate to the mc-hash crate, where it's better grouped with its domain

AmbientTea commented on primitives/sidechain-mc-hash/src/lib.rs:
```diff
@@ -201,6 +207,38 @@ impl InherentDigest for McHashInherentDigest {
 	}
 }
 
+pub fn inherent_digest_value_for_hash<ID: InherentDigest, Block: BlockT, C>(
```
This is something that should ideally reside with the `InherentDigest` trait itself, but I added this here for now

LGLO replied:
1) Regarding your comment: I'm not sure if this is generic enough. Why for block number zero there is None returned? Maybe some implementations could get it out from genesis?

2) method name suggests it can get digest from hash, I suggest `inherent_digest_from_block`.

AmbientTea replied:
Genesis block has an empty digest log and there's no mechanism that can put an inherent digest there.

LGLO commented on node/src/tests/inherent_data_tests.rs:
```diff
@@ -20,24 +22,28 @@ async fn block_proposal_cidp_should_be_created_correctly() {
 
 	#[cfg(not(feature = "block-beneficiary"))]
-	let (slot, timestamp, mc_hash, ariadne_data) = inherent_data_providers;
+	let (slot, timestamp, mc_hash, ariadne_data, native_token) = inherent_data_providers;
```
Side note: why are we having `block-beneficiary` feature?

AmbientTea replied:
I think this is just a left-over. I guess we can remove it, I'll push a separate PR for that.

AmbientTea replied:
https://github.com/input-output-hk/partner-chains/pull/50

LGLO commented on runtime/src/lib.rs:
```diff
@@ -838,6 +838,16 @@ impl_runtime_apis! {
+				illiquid_supply_address: Default::default(),
+			}
+		}
+	}
```
I expect this dummy implementation to be changed soon, but already heads-up that we'll need a migration or migration guide at least for it.

LGLO commented on primitives/sidechain-mc-hash/src/lib.rs:
```diff
@@ -3,17 +3,19 @@ use main_chain_follower_api::{
 use sp_inherents::{InherentData, InherentDataProvider, InherentDigest, InherentIdentifier};
-use sp_runtime::{traits::Header as HeaderT, DigestItem};
+use sp_runtime::traits::Zero;
+use sp_runtime::{traits::Block as BlockT, traits::Header as HeaderT, DigestItem};
```
Cosmetic: please group them, now Zero is separate to fellow Block and Header

LGLO commented on primitives/sidechain-mc-hash/src/lib.rs:
```diff
@@ -119,6 +121,10 @@ impl McHashInherentDataProvider {
+
+	pub fn mc_hash(&self) -> McBlockHash {
+		self.mc_block.hash.clone()
+	}
```
Help my understand: how was `McBlockHash` extracted from  `McHashInherentDataProvider` before? I'm surprised that we accessed hash in a different way or we didn't at all.

LGLO commented on primitives/native-token-management/src/tests/runtime_api_mock.rs:
```diff
@@ -0,0 +1,78 @@
+		&self,
+		_number: NumberFor<Block>,
+	) -> Result<Option<<Block as BlockT>::Hash>, sp_blockchain::Error> {
+		unimplemented!()
```
`fn hash` is added in this PR and mock has it unimplemented - it indicates missing tests.

AmbientTea replied:
It's not the same hash. This is a `hash` function on `HeaderBackend`, which is not needed here. In fact, only `header()` is uses, so I changed the rest of them to `uninplemented!()` too.

LGLO commented on primitives/native-token-management/src/lib.rs:
```diff
@@ -0,0 +1,117 @@
+#[cfg(test)]
+mod tests;
+
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"ntreserv";
```
Please add a comment why `ntreserv`

AmbientTea replied:
It came from 'native token reserve', but it's not just connected to the reserve. Anywya, I renamed it to `nattoken` so it's clearer.
